feat(preflights): add arp filtering related preflights

JGAntunes opened the pull request:
#### What this PR does / why we need it:
~Keeping the PR in draft mode as we try to create `sysctl` specifc preflights instead:~
- https://github.com/replicatedhq/troubleshoot/issues/1674
- https://github.com/replicatedhq/troubleshoot/issues/1675

Changes have been merged, I'll work on releasing troubleshoot and testing the spec.

Docs for reference - https://www.kernel.org/doc/html/latest/networking/ip-sysctl.html

#### Which issue(s) this PR fixes:
https://app.shortcut.com/replicated/story/114999/add-preflights-for-arp-filter-and-arp-ignore-kernel-parameters

#### Does this PR require a test?
<!---
If no, just write "NONE" below.
-->
Simple spec change, but run a test locally:

```bash
root@node0:/replicatedhq/embedded-cluster# output/bin/embedded-cluster install run-preflights --no-prompt --license local-dev/Joao.yaml
✔  Host files materialized!
✔  Host preflights succeeded!
Host preflights completed successfully
root@node0:/replicatedhq/embedded-cluster# sysctl net.ipv4.conf.default.arp_filter=1
net.ipv4.conf.default.arp_filter = 1
root@node0:/replicatedhq/embedded-cluster# sysctl net.ipv4.conf.default.arp_ignore=1
net.ipv4.conf.default.arp_ignore = 1
root@node0:/replicatedhq/embedded-cluster# sysctl net.ipv4.conf.all.arp_filter=1
net.ipv4.conf.all.arp_filter = 1
root@node0:/replicatedhq/embedded-cluster# sysctl net.ipv4.conf.all.arp_ignore=1
net.ipv4.conf.all.arp_ignore = 1
root@node0:/replicatedhq/embedded-cluster# output/bin/embedded-cluster install run-preflights --no-prompt --license local-dev/Joao.yaml
✔  Host files materialized!
✗  4 host preflights failed

 •  ARP filtering is enabled by default for newly created interfaces on the host. Disable it by editing `/etc/sysctl.conf` and adding the line
    `net.ipv4.conf.default.arp_filter=0` followed by running `sudo sysctl -p`.
 •  ARP ignore is enabled by default for newly created interfaces on the host. Disable it by editing `/etc/sysctl.conf` and adding the line
    `net.ipv4.conf.default.arp_ignore=0` followed by running `sudo sysctl -p`.
 •  ARP filtering is enabled for all interfaces on the host. Disable it by editing `/etc/sysctl.conf` and adding the line
    `net.ipv4.conf.all.arp_filter=0` followed by running `sudo sysctl -p`.
 •  ARP ignore is enabled for all interfaces on the host. Disable it by editing `/etc/sysctl.conf` and adding the line
    `net.ipv4.conf.all.arp_ignore=0` followed by running `sudo sysctl -p`.

Please address these issues and try again.
```

#### Does this PR require a release note?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
-->
```release-note
Adds preflight checks for the `arp_ignore` and `arp_filter` kernel parameters
```

#### Does this PR require documentation?
NONE
<!--
If no, just write "NONE" below.
If yes, link to the related https://github.com/replicatedhq/replicated-docs documentation PR:
-->

ricardomaraschini reviewed (commented):
Shouldn't we use the collector introduced by https://github.com/replicatedhq/embedded-cluster/pull/1449 ?

JGAntunes commented:
@ricardomaraschini we're keeping this on pause while we look into:
- https://github.com/replicatedhq/troubleshoot/issues/1674
- https://github.com/replicatedhq/troubleshoot/issues/1675

But yeah, if we end up going with this PR we should use the collector you introduced.

ricardomaraschini commented:
> @ricardomaraschini we're keeping this on pause while we look into:
> 
> * [Create a `sysctl` host collector troubleshoot#1674](https://github.com/replicatedhq/troubleshoot/issues/1674)
> * [Create a `sysctl` host analyzer troubleshoot#1675](https://github.com/replicatedhq/troubleshoot/issues/1675)
> 
> But yeah, if we end up going with this PR we should use the collector you introduced.

Oh, nice. In the end we would migrate this (and mine) to use the new collector and analyzer. Thanks.

JGAntunes commented:
@ajp-io I've updated the spec, feel free to 👀 the preflight text and push any changes that you deem worth. I'll work on updating troubleshoot and test the spec next week.

ajp-io commented on pkg/preflights/host-preflight.yaml:
```diff
@@ -834,3 +835,18 @@ spec:
+    - sysctl:
+        checkName: Sysctl Kernel Parameters
+        outcomes:
+          - fail:
+              when: 'net.ipv4.conf.default.arp_filter > 0'
+              message: "ARP filtering is enabled by default for newly created interfaces on the host. Disable it by running 'sysctl net.ipv4.conf.default.arp_filter=0'."
+          - fail:
+              when: 'net.ipv4.conf.default.arp_ignore > 0'
+              message: "ARP ignore is enabled by default for newly created interfaces on the host. Disable it by running 'sysctl net.ipv4.conf.default.arp_ignore=0'."
+          - fail:
+              when: 'net.ipv4.conf.all.arp_filter > 0'
+              message: "ARP filtering is enabled for all interfaces on the host. Disable it by running 'sysctl net.ipv4.conf.all.arp_filter=0'."
+          - fail:
+              when: 'net.ipv4.conf.all.arp_ignore > 0'
+              message: "ARP ignore is enabled for all interfaces on the host. Disable it by running 'sysctl net.ipv4.conf.all.arp_ignore=0'."
```
I think you want a pass condition here too.

As I understand Troubleshoot, you hit a matching `when` and drop out there. So if someone has filter and ignore set incorrectly, they would only get the message that comes first in this list. Is that correct? If so, can we make this so that all failure conditions would show? Not sure if that needs to be two analyzers or something else.

JGAntunes replied:
You're right, I misunderstood `troubleshoot` in that aspect. Tested it out locally now and looks 👍

ajp-io replied:
The separate analyzers looks good. I still think you need a `pass` for each in addition to the fail. See the other analyzers in there, like [here](https://github.com/replicatedhq/embedded-cluster/blob/9db918c9d1014bf414da1edbd7abf28ffab64208/pkg/preflights/host-preflight.yaml#L831-L836).

ajp-io commented on pkg/preflights/host-preflight.yaml:
```diff
@@ -834,3 +835,18 @@ spec:
+        checkName: Sysctl Kernel Parameters
+        outcomes:
+          - fail:
+              when: 'net.ipv4.conf.default.arp_filter > 0'
```
If someone uses a network interface other than default, does this value need to use that interface instead?

JGAntunes replied:
So the `default` entry doesn't really translate into a particular interface but is used instead whenever a new interface is created, that's the value that gets set, which is exactly what happens when `calico` spins up its specific interfaces (and that's what poses a problem).

ajp-io commented on pkg/preflights/host-preflight.yaml:
```diff
@@ -834,3 +835,18 @@ spec:
+        outcomes:
+          - fail:
+              when: 'net.ipv4.conf.default.arp_filter > 0'
+              message: "ARP filtering is enabled by default for newly created interfaces on the host. Disable it by running 'sysctl net.ipv4.conf.default.arp_filter=0'."
```
I'm not sure these instructions make the change permanent. I think you might want to edit `/etc/sysctl.conf` and then use `sysctl -p`.

JGAntunes replied:
True, changed the instructions in - https://github.com/replicatedhq/embedded-cluster/pull/1454/commits/6c800ba25f43a182f43b61f63ca7fe8f64c71444

JGAntunes commented:
@ajp-io addressed your review and have run this locally:

```bash
root@node0:/replicatedhq/embedded-cluster# output/bin/embedded-cluster install run-preflights --no-prompt --license local-dev/Joao.yaml
✔  Host files materialized!
✔  Host preflights succeeded!
Host preflights completed successfully
root@node0:/replicatedhq/embedded-cluster# sysctl net.ipv4.conf.default.arp_filter=1
net.ipv4.conf.default.arp_filter = 1
root@node0:/replicatedhq/embedded-cluster# sysctl net.ipv4.conf.default.arp_ignore=1
net.ipv4.conf.default.arp_ignore = 1
root@node0:/replicatedhq/embedded-cluster# sysctl net.ipv4.conf.all.arp_filter=1
net.ipv4.conf.all.arp_filter = 1
root@node0:/replicatedhq/embedded-cluster# sysctl net.ipv4.conf.all.arp_ignore=1
net.ipv4.conf.all.arp_ignore = 1
root@node0:/replicatedhq/embedded-cluster# output/bin/embedded-cluster install run-preflights --no-prompt --license local-dev/Joao.yaml
✔  Host files materialized!
✗  4 host preflights failed

 •  ARP filtering is enabled by default for newly created interfaces on the host. Disable it by editing `/etc/sysctl.conf` and adding the line
    `net.ipv4.conf.default.arp_filter=0` followed by running `sudo sysctl -p`.
 •  ARP ignore is enabled by default for newly created interfaces on the host. Disable it by editing `/etc/sysctl.conf` and adding the line
    `net.ipv4.conf.default.arp_ignore=0` followed by running `sudo sysctl -p`.
 •  ARP filtering is enabled for all interfaces on the host. Disable it by editing `/etc/sysctl.conf` and adding the line
    `net.ipv4.conf.all.arp_filter=0` followed by running `sudo sysctl -p`.
 •  ARP ignore is enabled for all interfaces on the host. Disable it by editing `/etc/sysctl.conf` and adding the line
    `net.ipv4.conf.all.arp_ignore=0` followed by running `sudo sysctl -p`.

Please address these issues and try again.
```

ajp-io commented:
I left a comment on an outdated comment, which means you might not see it. Here it is https://github.com/replicatedhq/embedded-cluster/pull/1454#discussion_r1838652695

JGAntunes commented:
@ajp-io addressed via https://github.com/replicatedhq/embedded-cluster/pull/1454/commits/37354815ccc7786fe9c34fa11c1c5652bb0010d0

ajp-io commented:
lgth, though I don't know if you need an engineer's review too
